Rubberduck.Inspections/Concrete/ShadowedDeclarationInspection.cs

```diff
@@ -0,0 +1,628 @@
+        {
+            if (builtInEventHandlers.Contains(declaration))
+            {
+                return true;
+            }
+
+            var parameterDeclaration = declaration as ParameterDeclaration;
+
+            return parameterDeclaration != null && builtInEventHandlers.Contains(parameterDeclaration.ParentDeclaration);
+        }
+
+        private static bool DeclarationCanBeShadowed(Declaration originalDeclaration, Declaration userDeclaration, DeclarationSite originalDeclarationSite,
+            OptionPrivateModuleListener optionPrivateModuleListener, EnumerationRuleIndexListener enumerationRuleIndexListener)
+        {
+            if (originalDeclarationSite == DeclarationSite.NotApplicable)
+            {
+                return false;
+            }
+
+            if (originalDeclarationSite == DeclarationSite.ReferencedProject)
+            {
+                return DeclarationInReferencedProjectCanBeShadowed(originalDeclaration, userDeclaration, optionPrivateModuleListener);
+            }
+
+            if (originalDeclarationSite == DeclarationSite.OtherComponent)
+            {
+                return DeclarationInAnotherComponentCanBeShadowed(originalDeclaration, userDeclaration, optionPrivateModuleListener);
+            }
+
+            return DeclarationInTheSameComponentCanBeShadowed(originalDeclaration, userDeclaration, enumerationRuleIndexListener);
+        }
+
+        private static bool DeclarationInReferencedProjectCanBeShadowed(Declaration originalDeclaration, Declaration userDeclaration, OptionPrivateModuleListener listener)
+        {
+            if (DeclarationIsInsideOptionPrivateModule(originalDeclaration, listener))
+            {
+                return false;
+            }
+
+            var originalDeclarationComponentType = originalDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+            var userDeclarationComponentType = userDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+
+            // It is not possible to directly access a Parameter, UDT Member or Label declared in another project
+            if (originalDeclaration.DeclarationType == DeclarationType.Parameter || originalDeclaration.DeclarationType == DeclarationType.UserDefinedTypeMember ||
+                originalDeclaration.DeclarationType == DeclarationType.LineLabel)
+            {
+                return false;
+            }
+
+            // It is not possible to instantiate a Class Module which is not exposed
+            if ((originalDeclaration as ClassModuleDeclaration)?.IsExposed == false)
+            {
+                return false;
+            }
+
+            // It is not possible to directly access a UserForm or Document declared in another project, nor any declarations placed inside them
+            if (originalDeclarationComponentType == ComponentType.UserForm || originalDeclarationComponentType == ComponentType.Document)
```

Comment:
Hmm, I would check the document module case. In Access at least, it is possible to reference to referenced Access database's forms/ reports. I don't know if other hosts has similar behaviors.

Reply:
I'm no Access expert, so I might be misinterpreting your comment. But when I created a form in Access and saved it, it's not visible in the IDE. Manually I can only insert a new procedural module or class module into my VBA project. So unless I don't understand you correctly, this code doesn't impact Access at all.

Reply:
This might help clarify... [DoCmd.OpenForm](https://msdn.microsoft.com/en-us/vba/access-vba/articles/docmd-openform-method-access?f=255&MSPPError=-2147217396)...

In particular this tidbit:
     
    A string expression that's the valid name of a form in the current database. If you execute 
    Visual Basic code containing the OpenForm method in a library database, Microsoft 
    Access looks for the form with this name first in the library database, then in the current 
    database.
    
In this specific instance, it refers to using a string but given that you can instantiate an Access form using syntax such as `Dim frm As Form_frmMyCustomForm`, this could potentially be shadowing if the library database and the main database both contains a form named `frmMyCustomForm`.

Reply:
@bclothier I'm not really worried about false negatives; that's good to know though - actually, hold on a second - does the inspection pick up shadowed *types* (as is the case with this example)? I don't think it should. The fix for a shadowed identifier is to rename it; the fix for a shadowed type is to *qualify* it.

Reply:
I'll take a closer look a the code, but if it's currently issuing results for an ambiguous AsTypeClause, then I think it would be useful to include a piece of data in the "properties" parameter of the inspection result - e.g. "AsTypeName", so that the quickfix provider can eventually offer a "Qualify type" fix for these cases.

Reply:
@bclothier In the library database I have a form called Form1. In the user database I referenced the library database. The line `Dim f As Form1` results in a "User-defined type not defined" error. The line `DoCmd.OpenForm "Form1"` results in a "The form name Form1 is misspelled or refers to a form that doesn't exist" error.

Reply:
@retailcoder The inspection does nothing with AsType clauses. It only complains when there are clashing identifiers.

Reply:
@bclothier What's more, Rubberduck doesn't consider such form a declaration. Inspecting `State.UserDeclarations` doesn't return anything for the form. So since it's not a VBA declaration (at least not one picked up by Rubberduck), then we shouldn't worry about it.

Reply:
@rkapka After seeing your comments, i went and did some testing. You are correct - the form declarations is not shared across the projects. From the "main" project, the Library project can only expose the public procedure in VBA modules; forms are not exposed. 

However, `DoCmd.OpenForm` does work as expected; it's just that it's the other way around. E.g. if there's a form named `frmMain` that exists only in the main file but NOT in library database, then running `DoCmd.OpenForm "frmMain"` from a procedure within Library will succeed anyway, even though it's technically not possible to reference the form directly in the VBA code. 

Apologies for bringing up what turned out be a non-issue.

Reply:
@bclothier No need to apologize. I appreciate that someone invests their time to look into my code and think about possible oversights.

>it's just that it's the other way around. E.g. if there's a form named frmMain that exists only in the main file but NOT in library database, then running DoCmd.OpenForm "frmMain" from a procedure within Library will succeed

Now that's weird...